File: src/trending.js
// Trending / recency-aware ranking.
//
// Each query carries an exponentially-decayed "recent activity" counter:
//
//   on each search:   recent = recent * 2^(-Δt / halfLife) + 1 ;  ts = now
//   to read it:       effective = recent * 2^(-(now - ts) / halfLife)
//
// Properties this gives us (the points the assignment asks us to explain):
//
//  * How recent searches are tracked:  a single decayed float + timestamp per
//    active query — no need to store every individual search event.
//  * How recent activity affects ranking:  the blended score adds
//    recencyWeight * effectiveRecent on top of log(allTimeCount), so a query
//    being searched a lot right now climbs above an equally-or-more popular
//    all-time query.
//  * Why short-lived spikes don't dominate forever:  the decay is exponential,
//    so a burst loses half its weight every `halfLife` ms. A query that was hot
//    for five minutes is back near zero soon after — only sustained activity
//    keeps a high recent score. All-time count (monotonic) preserves long-term
//    popularity; recency only re-orders.
//
// Only queries with recent activity live in this map, so it stays small and the
// top-N scan for /trending is cheap. Negligible entries are pruned.

import { config } from './config.js';

const LN2 = Math.log(2);

export class Trending {
  constructor({
    halfLifeMs = config.trending.halfLifeMs,
    recencyWeight = config.trending.recencyWeight,
    maxTracked = 20_000,
    epsilon = 1e-3, // prune entries whose effective score falls below this
  } = {}) {
    this.halfLifeMs = halfLifeMs;
    this.recencyWeight = recencyWeight;
    this.maxTracked = maxTracked;
    this.epsilon = epsilon;
    this.map = new Map(); // query -> { recent, ts }
  }

  _decay(dtMs) {
    if (dtMs <= 0) return 1;
    return Math.pow(2, -dtMs / this.halfLifeMs);
  }

  // Register a search for `query`.
  record(query, now = Date.now()) {
    const e = this.map.get(query);
    if (!e) {
      this.map.set(query, { recent: 1, ts: now });
    } else {
      e.recent = e.recent * this._decay(now - e.ts) + 1;
      e.ts = now;
    }
    if (this.map.size > this.maxTracked) this.prune(now);
  }

  // Effective (decayed-to-now) recent score for a query, 0 if untracked.
  effectiveRecent(query, now = Date.now()) {
    const e = this.map.get(query);
    if (!e) return 0;
    return e.recent * this._decay(now - e.ts);
  }

  // Blended ranking score used by the enhanced suggestion ordering.
  // log10 compresses the Zipf-huge all-time counts so recency can meaningfully
  // reorder without a single mega-popular query swamping everything.
  blend(allTimeCount, effectiveRecent) {
    return Math.log10(allTimeCount + 1) + this.recencyWeight * effectiveRecent;
  }

  // Persisted recency state for the batch writer to flush to the DB.
  snapshotFor(query, now = Date.now()) {
    const e = this.map.get(query);
    if (!e) return { recentScore: 0, recentTs: now };
    return { recentScore: e.recent * this._decay(now - e.ts), recentTs: now };
  }

  // Seed recency state from the DB at startup (so trending survives a restart).
  loadPersisted(rows, now = Date.now()) {
    for (const r of rows) {
      if (r.recent_score && r.recent_score > this.epsilon && r.recent_ts) {
        this.map.set(r.query, { recent: r.recent_score, ts: r.recent_ts });
      }
    }
    this.prune(now);
  }

  // Top-N currently-trending queries by effective recent score.
  top(limit = config.trending.trendingLimit, now = Date.now()) {
    const arr = [];
    for (const [query, e] of this.map) {
      const eff = e.recent * this._decay(now - e.ts);
      if (eff > this.epsilon) arr.push({ query, recentScore: eff });
    }
    arr.sort((a, b) => b.recentScore - a.recentScore);
    return arr.slice(0, limit);
  }

  // Trending queries that start with `prefix` — injected into the suggestion
  // candidate pool so a surging query can surface even if its all-time count is
  // too low to be in the Trie's top-N-by-count pool.
  matching(prefix, limit, now = Date.now()) {
    const out = [];
    for (const [query, e] of this.map) {
      if (query.startsWith(prefix)) {
        const eff = e.recent * this._decay(now - e.ts);
        if (eff > this.epsilon) out.push({ query, recentScore: eff });
      }
    }
    out.sort((a, b) => b.recentScore - a.recentScore);
    return out.slice(0, limit);
  }

  // Drop negligible entries; if still over capacity, keep only the strongest.
  prune(now = Date.now()) {
    for (const [query, e] of this.map) {
      const eff = e.recent * this._decay(now - e.ts);
      if (eff <= this.epsilon) this.map.delete(query);
    }
    if (this.map.size > this.maxTracked) {
      const sorted = [...this.map.entries()].sort(
        (a, b) =>
          b[1].recent * this._decay(now - b[1].ts) -
          a[1].recent * this._decay(now - a[1].ts)
      );
      this.map = new Map(sorted.slice(0, this.maxTracked));
    }
  }

  get size() {
    return this.map.size;
  }
}

export default Trending;

File: app/trending.py
"""Recency-aware ranking via an exponentially-decayed activity counter.

Each query carries a `recent` score that decays with a configurable half-life.
On every search we first decay the old score to *now*, then add 1. Because the
score halves every `half_life` seconds, a burst of activity boosts a query
temporarily and then fades — so a one-day-popular query can't permanently
out-rank everything (the failure mode of a raw lifetime counter).

Two consumers:
  - mode=recency suggestions blend it with the historical count:
        score = log(count + 1) + recency_weight * decayed_recent
    The log compresses the huge spread of all-time counts so a genuine recent
    surge can actually move a suggestion up, without letting recency erase the
    signal that some queries are simply always popular.
  - GET /trending ranks purely by decayed recent activity.

When a flush changes activity, the recency-mode cache entries for the affected
prefixes are stale, so on_flush() invalidates exactly those Redis keys (plus the
global trending key).
"""
from __future__ import annotations

import math
import threading
import time
from typing import List, Optional

from app.config import config
from app.trie import Entry


class TrendingTracker:
    def __init__(self, cache=None, half_life_s: Optional[float] = None,
                 recency_weight: Optional[float] = None):
        self.cache = cache  # DistributedCache, for invalidation
        self.half_life_s = half_life_s if half_life_s is not None else config.trend_half_life_s
        self.recency_weight = recency_weight if recency_weight is not None else config.recency_weight
        self._decay_per_s = math.log(2) / self.half_life_s
        self._active: set[Entry] = set()
        self._lock = threading.Lock()

    def _now(self) -> float:
        return time.time()

    # ---- decay math ----
    def decayed(self, entry: Entry, now: Optional[float] = None) -> float:
        """Current decayed recent score (does not mutate the entry)."""
        if entry.recent <= 0.0:
            return 0.0
        now = now if now is not None else self._now()
        dt = now - entry.last_ts
        if dt <= 0:
            return entry.recent
        return entry.recent * math.exp(-self._decay_per_s * dt)

    def blended_score(self, entry: Entry, now: Optional[float] = None) -> float:
        return math.log(entry.count + 1) + self.recency_weight * self.decayed(entry, now)

    # ---- hot path: record one search (called from BatchWriter.record) ----
    def record(self, entry: Entry) -> None:
        now = self._now()
        entry.recent = self.decayed(entry, now) + 1.0
        entry.last_ts = now
        with self._lock:
            self._active.add(entry)

    # ---- trending list ----
    def trending(self, limit: Optional[int] = None) -> List[dict]:
        limit = limit or config.trending_limit
        now = self._now()
        with self._lock:
            active = list(self._active)
        scored = []
        survivors = []
        for e in active:
            d = self.decayed(e, now)
            if d < 1e-3:
                continue  # faded out — drop from the active set
            survivors.append(e)
            scored.append((e, d))
        # prune faded entries so the active set stays bounded
        if len(survivors) != len(active):
            with self._lock:
                self._active = set(survivors) | (self._active - set(active))
        scored.sort(key=lambda t: t[1], reverse=True)
        return [
            {"query": e.query, "recent": round(d, 3), "count": e.count,
             "score": round(self.blended_score(e, now), 4)}
            for e, d in scored[:limit]
        ]

    # ---- cache invalidation on flush ----
    def on_flush(self, snapshot: dict[str, int]) -> int:
        """Invalidate recency-mode cache for prefixes whose ranking may have moved.

        Returns the number of Redis keys invalidated. A flush means recent
        activity changed, so recency-mode results for any prefix of a searched
        query are potentially stale.
        """
        if self.cache is None:
            return 0
        prefixes: set[str] = set()
        for q in snapshot:
            for i in range(1, len(q) + 1):
                prefixes.add(q[:i])
        invalidated = 0
        for p in prefixes:
            self.cache.delete(self.cache.key_for(p, "recency"))
            invalidated += 1
        # global trending list is recomputed each flush
        self.cache.delete("trending:global")
        invalidated += 1
        return invalidated
